src/ethereum/osaka/bal_builder.py

```diff
@@ -0,0 +1,152 @@
+"""
+Block Access List Builder for EIP-7928
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+This module implements the BAL builder that tracks all account and storage
+accesses during block execution and constructs the final BlockAccessList.
+"""
+
+from collections import defaultdict
+from typing import Dict, Set
+
+from ethereum_types.bytes import Bytes
+from ethereum_types.numeric import U32, U64, Uint
+
+from .fork_types import Address
+from .ssz_types import (
+    AccountChanges,
+    BalanceChange,
+    BlockAccessList,
+    CodeChange,
+    NonceChange,
+    SlotChanges,
+    StorageChange,
+)
+
+
+class BALBuilder:
+    """
+    Builder for constructing BlockAccessList efficiently during transaction execution.
+    
+    Follows the pattern: address -> field -> tx_index -> change
+    """
+
+    def __init__(self) -> None:
+        # address -> field_type -> changes
+        self.accounts: Dict[Address, Dict[str, any]] = {}
+
+    def _ensure_account(self, address: Address) -> None:
```

Comment:
We try to keep our classes as dumb as possible (basically structs) and use unbound methods instead.